File: services/internal-api/start.py
#!/usr/bin/env python3
"""
Railway startup script for internal-api service.

This script handles PORT resolution and starts uvicorn with the correct settings.
It ensures environment variables are properly expanded before being passed to uvicorn.
"""

import os
import sys
import subprocess
import logging

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    handlers=[logging.StreamHandler()]
)
logger = logging.getLogger("startup")

def get_port():
    """Get PORT from environment or use default."""
    try:
        port_str = os.environ.get("PORT")
        logger.info(f"Raw PORT environment variable: {port_str!r}")
        
        if port_str is None:
            logger.info("PORT not set, using default 8000")
            return 8000
            
        try:
            port = int(port_str)
            logger.info(f"Using PORT: {port}")
            return port
        except ValueError:
            logger.error(f"Invalid PORT value: {port_str!r}, using default 8000")
            return 8000
    except Exception as e:
        logger.exception("Error resolving PORT")
        return 8000

def main():
    """Main entry point."""
    try:
        port = get_port()
        
        # Log all environment variables for debugging
        logger.info("Environment variables:")
        for key, value in os.environ.items():
            logger.info(f"  {key}={value}")
        
        # Build the uvicorn command
        cmd = [
            "uvicorn",
            "app.main:app",
            "--host", "0.0.0.0",
            "--port", str(port)
        ]
        
        logger.info(f"Starting uvicorn with command: {' '.join(cmd)}")
        
        # Execute uvicorn
        os.execvp("uvicorn", cmd)
    except Exception as e:
        logger.exception("Failed to start uvicorn")
        sys.exit(1)

if __name__ == "__main__":
    logger.info("Starting internal-api service...")
    main()

File: services/frontend/start.js
#!/usr/bin/env node
/**
 * Railway startup script for frontend service.
 * 
 * This script handles PORT resolution and starts the static file server
 * with the correct settings. It ensures environment variables are properly
 * expanded before being passed to the serve command.
 */

const { spawn } = require('child_process');
const path = require('path');

// Configure logging
function log(message) {
  const timestamp = new Date().toISOString();
  console.log(`${timestamp} - startup - ${message}`);
}

function getPort() {
  try {
    const portStr = process.env.PORT;
    log(`Raw PORT environment variable: ${JSON.stringify(portStr)}`);
    
    if (portStr === undefined || portStr === null) {
      log('PORT not set, using default 3000');
      return 3000;
    }
    
    try {
      const port = parseInt(portStr, 10);
      if (isNaN(port)) {
        log(`Invalid PORT value: ${JSON.stringify(portStr)}, using default 3000`);
        return 3000;
      }
      log(`Using PORT: ${port}`);
      return port;
    } catch (err) {
      log(`Error parsing PORT: ${err.message}, using default 3000`);
      return 3000;
    }
  } catch (err) {
    log(`Exception resolving PORT: ${err.message}`);
    return 3000;
  }
}

function main() {
  try {
    const port = getPort();
    
    // Log all environment variables for debugging
    log('Environment variables:');
    Object.entries(process.env).forEach(([key, value]) => {
      log(`  ${key}=${value}`);
    });
    
    // Build the serve command
    const serveArgs = ['-s', 'build', '-l', port.toString()];
    
    log(`Starting serve with command: serve ${serveArgs.join(' ')}`);
    
    // Execute serve using spawn
    const serveProcess = spawn('serve', serveArgs, {
      stdio: 'inherit'
    });
    
    serveProcess.on('error', (err) => {
      log(`Failed to start serve: ${err.message}`);
      process.exit(1);
    });
    
    serveProcess.on('close', (code) => {
      if (code !== 0) {
        log(`serve exited with code ${code}`);
        process.exit(code);
      }
    });
  } catch (err) {
    log(`Failed to start serve: ${err.message}`);
    process.exit(1);
  }
}

log('Starting frontend service...');
main();
